README.MD
[![Build Status](https://travis-ci.org/ttrinh/react-inlinese.svg?branch=master)](https://travis-ci.org/ttrinh/react-inlinese)

## Demo & Documentation
> [https://ttrinh.github.io/react-inlinese](https://ttrinh.github.io/react-inlinese/)

![alt text](https://ttrinh.github.io/react-inlinese/assets/react-inlinese.gif "react-inlinese")

## Installation
- npm
```
npm install --save react-inlinese
```

- OR yarn
```
yarn add react-inlinese
```

## Usage
```js
import Inlinese from 'react-inlinese';

// ...
// see documentation for available options
const MyComponent = () => (
    <Inlinese
        onSubmit={value => alert(value)}
        value="Editable text"
    >
        Editable text
    </Inlinese>
);

```

CHANGELOG.MD
## [v1.0.0] - 2017-12-18
### Change
- Apply & Cancel buttons' positions switched.
- Simplify theme and styles.
    + Remove prop "style" as it's inflexible to style children.

### Bug Fix
- Text in the middle of a paragraph causes miscalculation on input box width.

test/.setup.js
require('babel-register');
require('./.color-fix');
require('jsdom-global')();

test/inline-editable.spec.js
import React from "react";
import { shallow } from "enzyme";
import { expect, assert } from "chai";

import Buttons from "../src/Buttons";
import InlineEditable from "../src/index";
import { Container, InputBox, Label, Hint } from "../src/style";
import {
  calcInputBoxStyle,
  findParentByClass,
  getAutoGrowStyle
} from "../src/helpers";

// generate fake key event
const pressKeyEvent = keyCode => ({
  keyCode,
  target: {
    value: String.fromCharCode(
      96 <= keyCode && keyCode <= 105 ? keyCode - 48 : keyCode
    )
  }
});

// no operation
const noop = () => {};

describe("Inline Editable ", () => {
  const Wrapper = (props, children = "test") =>
    shallow(
      <InlineEditable value="basic text" onSubmit={noop} {...props}>
        {children}
      </InlineEditable>
    );

  it("it should not render without value and onSubmit", () => {
    const wrapper = shallow(<InlineEditable />);
    const wrapperNoValue = shallow(<InlineEditable onSubmit={noop} />);
    const wrapperNoSubmit = shallow(<InlineEditable value="one" />);

    expect(wrapper.children()).to.have.length(0);
    expect(wrapperNoValue.children()).to.have.length(0);
    expect(wrapperNoSubmit.children()).to.have.length(0);
  });

  it("it should render with value and onSubmit", () => {
    expect(Wrapper().children()).to.have.length(2);
  });

  it('it should assign "value" in state based off of value prop', () => {
    expect(Wrapper().state().value).to.equal("basic text");
  });

  it('it should render the children in Label, and add "rie" class.', () => {
    const childrenBox = Wrapper().find(Label);
    expect(childrenBox).to.have.length(1);
    expect(childrenBox.hasClass("rie")).to.equal(true);
    expect(childrenBox.containsMatchingElement("test")).to.equal(true);
  });

  it("it should not render Edit Icon when showEditIcon disabled", () => {
    expect(Wrapper().find(".rie-edit-indicator")).to.have.length(1);
    const NoEditIcon = Wrapper({ showEditIcon: false });
    expect(NoEditIcon.find(".rie-edit-indicator")).to.have.length(0);
  });

  it("it should not render Buttons when showButtons set to false", () => {
    const W = Wrapper();
    W.setState({ show: true });
    expect(W.find(Buttons)).to.have.length(1);

    const NoButtons = Wrapper({ showButtons: false });
    NoButtons.setState({ show: true });
    expect(NoButtons.find(Buttons)).to.have.length(0);
  });

  it("it should change state.show to show input box when the text is clicked ", () => {
    const W = Wrapper();
    W.find(Label).simulate("click");
    expect(W.state().show).to.equal(true);
  });

  it("it should change state.show to show input box when the text is focused ", () => {
    const W = Wrapper();
    W.find(Label).simulate("focus");
    expect(W.state().show).to.equal(true);
  });

  it('it should show the input box by adding "rie-show" class', () => {
    const W = Wrapper();
    W.find(Label).simulate("click");
    expect(W.find(InputBox).hasClass("rie-show")).to.equal(true);
  });

  it("it should change value when type", () => {
    const W = Wrapper();
    W.setState({ show: true });
    const Textarea = W.find("textarea");
    Textarea.simulate("change", { target: { value: "hello friends" } });
    expect(W.state().value).to.equal("hello friends");
  });

  it("it should submit value on Enter", () => {
    const W = Wrapper();
    W.setState({ show: true });
    const Textarea = W.find("textarea");

    Textarea.simulate("change", { target: { value: "hello friends" } });
    Textarea.simulate("keyDown", pressKeyEvent(13));
    expect(W.state().value).to.equal("hello friends");
    expect(W.state().show).to.equal(false);
  });

  it("it should format text with user defined formatter", () => {
    // no dollar sign formatter
    const noDollarFormatter = value =>
      value.match(/\$/) ? value.replace(/\$/g, "") : value;

    const W = Wrapper({
      formatter: noDollarFormatter
    });
    W.setState({ show: true });
    const Textarea = W.find("textarea");
    Textarea.simulate("change", { target: { value: "hello$" } });
    expect(W.state().value).to.equal("hello");
  });

  it("it should reset value on Esc or Tab", () => {
    const W = Wrapper();
    W.setState({ show: true });
    const Textarea = W.find("textarea");

    Textarea.simulate("change", { target: { value: "hello friends" } });
    Textarea.simulate("keyDown", pressKeyEvent(27));
    expect(W.state().value).to.equal("basic text");
    expect(W.state().show).to.equal(false);

    W.setState({ show: true });
    Textarea.simulate("change", { target: { value: "hello friends" } });
    Textarea.simulate("keyDown", pressKeyEvent(9));
    expect(W.state().value).to.equal("basic text");
    expect(W.state().show).to.equal(false);
  });

  it("it should trim value on submit", () => {
    const W = Wrapper();
    W.setState({ show: true });
    const Textarea = W.find("textarea");

    Textarea.simulate("change", { target: { value: "   hello friends    " } });
    Textarea.simulate("keyDown", pressKeyEvent(13));
    expect(W.state().value).to.equal("hello friends");
  });

  it("it should pass down hoverStyleString, default or user-define", () => {
    const def = "border-bottom: 1px dashed rgba(0,0,0,.2);";
    expect(Wrapper().instance().props.hoverStyleString).to.equal(def);
    const W = Wrapper({ hoverStyleString: "color: white" }).instance();
    expect(W.props.hoverStyleString).to.equal("color: white");
  });

  it("it should pass down placeholder, default or user-define", () => {
    expect(Wrapper().instance().props.placeholder).to.equal("");
    const W = Wrapper({ placeholder: "hello" }).instance();
    expect(W.props.placeholder).to.equal("hello");
  });

  it('it should, on disabled, set hoverStringStyle to "cursor: inherit;"', () => {
    const W = Wrapper({ disabled: true })
      .find(Label)
      .shallow()
      .instance();
    expect(W.props.hoverStyleString).to.equal("cursor: inherit;");
  });

  it("it should, on disabled, hide the edit icon indicator", () => {
    const W = Wrapper();
    expect(W.find(".rie-edit-indicator").length).to.equal(1);
    const W2 = Wrapper({ disabled: true });
    expect(W2.find(".rie-edit-indicator").length).to.equal(0);
  });

  it("it should, on disabled, not switch to the input box", () => {
    const W = Wrapper({ disabled: true });
    W.find(Label).simulate("click");
    expect(W.state().show).to.equal(false);
  });
});

describe("Box Helpers: calcInputBoxStyle", () => {
  // fake node
  const node = (w, h) =>
    (function() {
      function Node() {
        this.offsetWidth = w || 10;
        this.offsetHeight = h || 10;
        this.children = [];
      }

      Node.prototype.getBoundingClientRect = () => ({
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        width: w,
        height: h
      });

      return new Node();
    })();

  it("should return object", () => {
    assert.isObject(calcInputBoxStyle(node()));
  });

  it("should apply minWidth and minHeight", () => {
    const expected = { width: 300, height: 50 };
    expect(calcInputBoxStyle(node())).to.eql(expected);
  });

  it("should calculate width and height", () => {
    const newNode = node(300, 50);
    const expected = { width: 330, height: 90 };
    expect(calcInputBoxStyle(newNode)).to.eql(expected);
  });

  it("should calculate with a defined config", () => {
    const newNode = node(300, 50);
    const expected = { width: 390, height: 190 };
    const config = {
      padding: 20,
      extraSpace: 50,
      minWidth: 200,
      minHeight: 50
    };
    expect(calcInputBoxStyle(newNode, config)).to.eql(expected);
  });

  it("should calculate autogrow height", () => {
    const newNode = node(300, 80);
    const currentStyle = {
      width: 100,
      height: 50
    };
    const expected = { width: 100, height: 100 };
    expect(getAutoGrowStyle(newNode, currentStyle)).to.eql(expected);
  });

  it("should not calculate autogrow height when both node heights are equal", () => {
    const newNode = node(300, 50);
    const currentStyle = {
      width: 100,
      height: 50
    };
    expect(getAutoGrowStyle(newNode, currentStyle)).to.eql(currentStyle);
  });
});

describe("Buttons", () => {
  const buttons = props =>
    shallow(<Buttons submit={noop} cancel={noop} {...props} />);

  it("should render default values", () => {
    const B = buttons();
    expect(B.props().bgColor).to.equal("#555");
    expect(B.props().textColor).to.equal("white");
    expect(B.props().roundness).to.equal("3px");
    expect(B.containsMatchingElement("submit")).to.equal(true);
    expect(B.containsMatchingElement("cancel")).to.equal(true);
  });

  it("should render defined values", () => {
    const B = buttons({
      submitText: "Taco",
      cancelText: "Kitty",
      bgColor: "yellow",
      textColor: "red",
      roundness: "0"
    });
    expect(B.props().bgColor).to.equal("yellow");
    expect(B.props().textColor).to.equal("red");
    expect(B.props().roundness).to.equal("0");
    expect(B.containsMatchingElement("Taco")).to.equal(true);
    expect(B.containsMatchingElement("Kitty")).to.equal(true);
  });
});


src/style.js
import styled, { keyframes } from "styled-components";

export const Container = styled.span`
  position: relative;
  text-align: left;
  cursor: inherit;

  box-sizing: border-box;
  * {
    box-sizing: border-box;
  }

  .rie {
  }

  .rie-edit-indicator-wrapper {
    position: relative;
  }

  .rie-edit-indicator {
    position: absolute;
    background-color: white;
    top: 0;
    right: 0;
    width: 24px;
    height: 24px;
    text-align: center;
    font-family: sans-serif;
    font-weight: normal;
    border-radius: 50%;
    border: 1px solid rgba(0, 0, 0, 0.2);
    transition: all 0.12s ease;
    color: #555;
    opacity: 0;
    line-height: 0;
    display: flex;
    align-items: center;
    justify-content: center;
  }

  &:hover .rie-edit-indicator {
    right: -24px;
    opacity: 1;
  }

  ${props => props.styleString};
`;

const elasticShow = keyframes`
    from { opacity: 0;  transform: scale(1, .7); height: 0px;}
    60% { opacity: .3; transform: scale(1, 1.1); height: auto;}
    100% { opacity: 1; transform: scale(1, 1); }
`;

export const InputBox = styled.span`
  border-radius: ${props => props.roundness};
  height: 0;
  opacity: 0;
  display: none;

  z-index: 200;
  position: absolute;
  top: -10px;
  left: -10px;
  overflow: hidden;
  background-color: white;
  box-shadow: rgba(0, 0, 0, 0.09) 0px 1px 6px, rgba(0, 0, 0, 0.09) 0px 1px 4px;
  border: 1px solid rgba(0, 0, 0, 0.1);
  border-color: ${props => props.color};

  &.rie-show {
    height: auto;
    display: block;
    opacity: 1;
    animation: ${elasticShow} 0.2s ease-in;
  }

  textarea,
  .textarea-clone {
    box-sizing: border-box;
    resize: none;
    outline: none;
    vertical-align: top;
    font: inherit;
    padding: 10px;
    transition: all 0.1s ease;
    margin: 0;
    overflow: hidden;
  }

  textarea {
    border: 0;
  }

  textarea:focus {
  }

  .textarea-clone {
    white-space: pre-wrap;
    visibility: hidden;
    height: auto !important;
    width: 100% !important;
    position: absolute;
    top: 0;
    left: 0;
    display: block;
    opacity: 0.8;
  }

  .rie-button-area {
    position: relative;
    line-height: 0;
    border-top: 1px dashed rgba(0, 0, 0, 0.1);
    padding: 5px;
    background-color: #f7f7f7;
    font-size: 14px;
  }
`;

export const Label = styled.span`
  cursor: pointer;
  position: relative;

  &:hover {
    ${props => props.hoverStyleString};
  }
`;

export const Hint = styled.span`
  color: ${props => props.color};

  position: absolute;
  user-select: none;
  left: 0;
  bottom: 0px;
  padding: 0.5em 1em;
  opacity: 0.4;
  font: bold 8px Arial, sans-serif;
  text-transform: uppercase;
  letter-spacing: -0.25px;

  span {
    margin-right: 5px;
  }

  b {
    display: inline-block;
    padding: 2px 3px;
    margin-right: 2px;
    border-radius: 3px;
    border: 1px solid rgba(0, 0, 0, 0.2);
  }

  ${props => props.styleString};
`;


src/Buttons.js
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import MdCheck from 'react-icons/lib/md/check';
import MdClear from 'react-icons/lib/md/clear';

const Container = styled.span`
  display: block;
  text-align: right;
  color: #ccc;
  line-height: 0;

  .rie-button {
    padding-left: 6px;
  }

  button {
    cursor: pointer;
    transition: all 0.2s ease;
    outline: none;
    padding: 4px 10px;

    /* color: ${props => props.textColor}; */
    /* background-color: ${props => props.bgColor}; */
    background-color: white;
    color: rgba(0, 0, 0, 0.8);
    border-radius: ${props => props.roundness};
    border: 1px solid rgba(0, 0, 0, 0.2);
    box-shadow: 0 0px 0px rgba(0,0,0,0.1), 0 1px 1px rgba(0,0,0,0.05);
    transition: all 0.3s ease-out);

    &:active {
      transform: scale(0.98, 0.98);
    }
  }

  button:hover {
    box-shadow: 0 0px 0px rgba(0,0,0,0.1), 0 1px 2px rgba(0,0,0,0.2);
  }
`;

const Buttons = ({
  submit,
  cancel,
  submitText,
  cancelText,
  bgColor,
  textColor,
  roundness
}) => (
  <Container bgColor={bgColor} textColor={textColor} roundness={roundness}>
    <span className="rie-button">
      <button className="submit" onClick={submit} title={submitText || 'apply'}>
        <MdCheck size={16} /> {submitText}
      </button>
    </span>

    <span className="rie-button">
      <button className="cancel" onClick={cancel} title={cancelText}>
        <MdClear size={16} /> {cancelText}
      </button>
    </span>
  </Container>
);

Buttons.propTypes = {
  submit: PropTypes.func.isRequired,
  cancel: PropTypes.func.isRequired,
  bgColor: PropTypes.string,
  textColor: PropTypes.string,
  roundness: PropTypes.string,
  submitText: PropTypes.string,
  cancelText: PropTypes.string
};

Buttons.defaultProps = {
  submitText: 'submit',
  cancelText: 'cancel',
  cancel: 'cancel',
  bgColor: '#555',
  textColor: 'white',
  roundness: '3px'
};

export default Buttons;


src/helpers.js
/**
* Find parent by className
*
* @param {Node} element     child element
* @param {String} className parent className to match
* @return {Node}            parent node or itself that has the class
**/
export const findParentByClass = (element, className) => (
    (element.className.split(' ').indexOf(className) >= 0)
    ? element
    : element.parentNode && findParentByClass(element.parentNode, className)
);


/**
* Calculate width and height of the input box
* based on width and height of the label text
*
* @param {Node}     node        the text label element
* @param {Object}   config      { padding, extraSpace, minWidth, minHeight }
* @return {Object}              { width, height }
**/
const defaultConfig = {
    padding: 10,
    extraSpace: 10,
    minWidth: 300,
    minHeight: 50,
};
export const calcInputBoxStyle = (node, config = defaultConfig) => {
    const { padding, extraSpace, minWidth, minHeight } = config;

    // The box should includes side paddings and extra space
    // vertical bottom padding should extend because the buttons & descriptions
    const sidePadding = padding * 2;
    const verticalExtra = extraSpace * 2;
    const w = node.offsetWidth + sidePadding + extraSpace;
    const h = node.offsetHeight + sidePadding + verticalExtra;

    // compare it against minimum values
    const width = w < minWidth ? minWidth : w;
    const height = h < minHeight ? minHeight : h;

    return {
        width,
        height,
    };
};


/**
* Get textarea auto-grow height throught its clone
*
* @param {Node}     cloneNode               node
* @param {Object}   currentStyle            textarea current style
* @return {Number}                          textarea height
**/
export const getAutoGrowStyle = (cloneNode, currentStyle) => {
    if (!cloneNode) return currentStyle;

    const cloneDimension = cloneNode.getBoundingClientRect();
    const cloneHeight = cloneDimension.height;
    const verticalExtra = defaultConfig.extraSpace * 2;

    // no grow when both heights are equal
    if (cloneHeight === currentStyle.height) {
        return currentStyle;
    }

    return {
        ...currentStyle,
        height: cloneHeight + verticalExtra,
    };
};



styleguide.config.js
const loaders = require('loaders');


module.exports = {
    components: 'src/index.js',
    webpackConfig: {
        module: {
            loaders: loaders.all,
        },
    },
    showSidebar: false,
    showCode: true,
    showUsage: true,
    styleguideDir: 'docs',
    highlightTheme: 'monokai',
    theme: {
        color: {
            base: '#333',
            light: '#999',
            lightest: '#ccc',
            link: '#1978c8',
            linkHover: '#f28a25',
            border: '#e8e8e8',
            name: '#4052b3',
            type: '#c3451c',
            error: '#fff',
            baseBackground: '#fff',
            errorBackground: '#c00',
            codeBackground: '#f5f5f5',
            sidebarBackground: '#f5f5f5',
        },
        fontFamily: {
            base: [
                '-apple-system',
                'BlinkMacSystemFont',
                '"Segoe UI"',
                '"Roboto"',
                '"Oxygen"',
                '"Ubuntu"',
                '"Cantarell"',
                '"Fira Sans"',
                '"Droid Sans"',
                '"Helvetica Neue"',
                'sans-serif',
            ],
            monospace: ['Consolas', '"Liberation Mono"', 'Menlo', 'monospace'],
        },
        fontSize: {
            base: 15,
            text: 16,
            small: 13,
            h1: 48,
            h2: 36,
            h3: 24,
            h4: 18,
            h5: 16,
            h6: 16,
        },
    },
    styles: {
        Playground: {
            preview: {
                borderRadius: 2,
                border: '1px solid rgba(0,0,0,0.3)',
                padding: '1em',
                width: '100%',
                // overflow: 'auto',
            },
        },
        Markdown: {
            pre: {
                // border: 0,
            },
            code: {
                fontSize: 14,
            },
        },
    },
};


src/index.js
import React from 'react';
import PropTypes from 'prop-types';
import MdCreate from 'react-icons/lib/md/create';

import Buttons from './Buttons';
import { Container, InputBox, Label, Hint } from './style';
import {
  findParentByClass,
  calcInputBoxStyle,
  getAutoGrowStyle
} from './helpers';

/**
 * React component for editing text inline.
 *
 * ###### Featuresconsole.log("dfasf");
 * - Simple & clean.
 * - Use Textarea for input.
 * - Autogrow text input height
 * - Format-able text input.
 * - Keyboard & Tab friendly).
 * - Simple theme customization.
 * - Support presentational element as String or Component. However, value must be string.
 */
class ReactInlinese extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.value || '',
      show: false,
      inputStyle: {}
    };

    this.switch = this.switch.bind(this);
    this.submit = this.submit.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.clickOutside = this.clickOutside.bind(this);
    this.autoGrowHeight = this.autoGrowHeight.bind(this);
    this.stopListeningOutclick = this.stopListeningOutclick.bind(this);
    this.startListeningOutclick = this.startListeningOutclick.bind(this);
  }

  componentWillUnmount() {
    this.stopListeningOutclick();
  }

  onChange(e) {
    const value = e.target.value;

    if (this.props.formatter) {
      const formattedValue = this.props.formatter.call(null, value);

      if (typeof formattedValue !== 'string') {
        console.warn(
          'ReactInlinese: formatter function must return a string value'
        );
        return;
      }

      this.setState({ value: formattedValue });
      this.autoGrowHeight();
      return;
    }

    this.setState({ value });
    this.autoGrowHeight();
  }

  onKeyDown(e) {
    if (e.keyCode === 13 && !e.shiftKey) this.submit(); // enter
    // escape or tab
    if (e.keyCode === 27 || e.keyCode === 9) {
      this.setState({ value: this.props.value }); // reset
      this.switch();
    }
  }

  // calculate the input styles.
  setInputStyle(element) {
    if (this.state.show) return;
    const node = findParentByClass(element, 'rie');
    this.setState({
      inputStyle: calcInputBoxStyle(node)
    });
  }

  autoGrowHeight() {
    const textareaAutoGrowStyle = getAutoGrowStyle(
      this.textareaClone,
      this.state.inputStyle
    );
    this.setState({ inputStyle: textareaAutoGrowStyle });
  }

  startListeningOutclick() {
    document.addEventListener('click', this.clickOutside, false);
  }
  stopListeningOutclick() {
    document.removeEventListener('click', this.clickOutside, false);
  }

  clickOutside(e) {
    const isInside = this.node.contains(e.target);
    if (!isInside) this.switch();
  }

  switch(e) {
    if (this.props.disabled) return;

    // Entering edit mode.
    // Add dom listener for click outside
    // Calculate Input Box style and set value
    if (!this.state.show) {
      this.startListeningOutclick();

      if (e) {
        this.setInputStyle(e.target);
        this.setState({ value: this.props.value });
      }

      // Not in edit mode. Remove the listener
    } else {
      this.stopListeningOutclick();
    }

    this.setState({ show: !this.state.show });
  }

  submit() {
    // no change
    if (this.state.value === this.props.value) {
      this.switch();
      return;
    }

    this.setState({ value: this.state.value.trim() });
    this.props.onSubmit(this.state.value);
    this.switch();
  }

  render() {
    const {
      value,
      onSubmit,
      disabled,
      submitText,
      cancelText,
      placeholder,
      primaryColor,
      secondaryColor,
      roundness,
      hoverStyleString,
      showButtons,
      showEditIcon,
      children,
      styles
    } = this.props;

    if (typeof value !== 'string' || !onSubmit) {
      return null;
    }

    return (
      <Container
        styleString={styles.main}
        innerRef={node => {
          this.node = node;
        }}
      >
        <Label
          className="rie"
          onClick={this.switch}
          onFocus={this.switch}
          tabIndex={disabled ? '' : '0'}
          onMouseEnter={this.hover}
          hoverStyleString={disabled ? 'cursor: inherit;' : hoverStyleString}
        >
          <span>
            {children}

            {!disabled &&
              showEditIcon &&
              !this.state.show && (
                <span className="rie-edit-indicator-wrapper">
                  <span className="rie-edit-indicator">
                    <MdCreate size={14} />
                  </span>
                </span>
              )}
          </span>
        </Label>

        <InputBox
          className={this.state.show ? 'rie-show' : ''}
          color={primaryColor}
          roundness={roundness}
        >
          {this.state.show && (
            <span
              className="textarea-clone"
              style={this.state.inputStyle}
              ref={node => {
                this.textareaClone = node;
              }}
            >
              {this.state.value}
            </span>
          )}
          {this.state.show && (
            <textarea
              tabIndex="-1"
              type="text"
              placeholder={placeholder}
              onKeyDown={this.onKeyDown}
              onChange={this.onChange}
              value={this.state.value}
              ref={input => input && input.focus()}
              style={this.state.inputStyle}
            />
          )}

          <div className="rie-button-area">
            {showButtons && (
              <Buttons
                submit={this.submit}
                cancel={this.switch}
                bgColor={primaryColor}
                textColor={secondaryColor}
                submitText={submitText}
                cancelText={cancelText}
                roundness={roundness}
              />
            )}

            <Hint
              color={primaryColor}
              title="Enter to apply, Esc to cancel"
              styleString={styles.hint}
            >
              <span>
                <b>Enter</b>Apply
              </span>
              <span>
                <b>Esc</b>Cancel
              </span>
            </Hint>
          </div>
        </InputBox>
      </Container>
    );
  }
}

ReactInlinese.propTypes = {
  /** Main value */
  value: PropTypes.string.isRequired,

  /** Process function when the text is submit */
  onSubmit: PropTypes.func.isRequired,

  /** Placeholder */
  placeholder: PropTypes.string,

  /** Disable editable */
  disabled: PropTypes.bool,

  /** Formatter. Acticate on value changed and pass down value as argument */
  formatter: PropTypes.func,

  /** Primary color. */
  primaryColor: PropTypes.string,

  /** Secondary color. Used for the buttons' text color */
  secondaryColor: PropTypes.string,

  /** Overall roundness, border-radius */
  roundness: PropTypes.string,

  /** Submit text for the input. */
  submitText: PropTypes.string,

  /** Cancel text for the input. */
  cancelText: PropTypes.string,

  /** Show edit indicator when the text is hovered. It's useful when the children element is a button or icon */
  showEditIcon: PropTypes.bool,

  /** Show Input Box buttons */
  showButtons: PropTypes.bool,

  /** STRING: Hover style */
  hoverStyleString: PropTypes.string,

  /** Representational display String(s) or Component(s) */
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.string,
    PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.element, PropTypes.string])
    )
  ]),

  /** STRING: Hover style */
  styles: PropTypes.shape({
    hint: PropTypes.string
  })
};

ReactInlinese.defaultProps = {
  submitText: 'apply',
  cancelText: 'cancel',
  placeholder: '',
  disabled: false,
  showEditIcon: true,
  showButtons: true,
  primaryColor: '#555',
  secondaryColor: 'white',
  roundness: '3px',
  hoverStyleString: 'border-bottom: 1px dashed rgba(0,0,0,.2);',
  styles: {
    main: '',
    hint: ''
  }
};

export default ReactInlinese;
